src/com/google/enterprise/gsafeed/GsafeedHelper.java

```diff
@@ -88,34 +92,77 @@ public void fatalError(SAXParseException exception)
           throw exception;
         }
       });
-    Document document = docBuilder.parse(inputStream);
-    return (Gsafeed) unmarshaller.unmarshal(document);
+    JAXBContext jaxbContext =
+        JAXBContext.newInstance("com.google.enterprise.gsafeed");
+    Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
+    UnmarshallerHandler unmarshallerHandler =
+        unmarshaller.getUnmarshallerHandler();
+    xmlReader.setContentHandler(unmarshallerHandler);
+    xmlReader.parse(new InputSource(inputStream));
+    return (Gsafeed) unmarshallerHandler.getResult();
   }
 
   /**
    * Avoid reading the DTD. No validation will happen.
    */
   public static Gsafeed unmarshalWithoutDtd(URL url) throws JAXBException,
       IOException, ParserConfigurationException, SAXException {
-    JAXBContext jaxbContext =
-        JAXBContext.newInstance("com.google.enterprise.gsafeed");
-    Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
-    DocumentBuilderFactory docBuilderFactory =
-        DocumentBuilderFactory.newInstance();
-    docBuilderFactory.setValidating(false);
-    DocumentBuilder docBuilder = docBuilderFactory.newDocumentBuilder();
-    docBuilder.setEntityResolver(new EntityResolver() {
+    return unmarshalWithoutDtd(url.openStream());
+  }
+
+  /**
+   * Avoid reading the DTD. No validation will happen.
+   */
+  public static Gsafeed unmarshalWithoutDtd(InputStream inputStream)
+      throws JAXBException, IOException, ParserConfigurationException,
+      SAXException {
+    SAXParserFactory saxParserFactory = SAXParserFactory.newInstance();
```

Comment:
Is there reasonably shareable code here, and in the unmarshaller, for helper methods? It seems like we are carefully constructing a secure parser, and code duplication seems like asking for trouble.

Reply:
Yes, but because you ask for changes to be as small as possible I left the refactoring for a later change. I thought of doing that possibly as part of updating XmlgroupsHelper, or even as a third step after XmlgroupsHelper has been updated and had a test suite added.

Reply:
SGTM